File: 2025-03-29/Copy_List_With_Random_Pointer.java
package amazon;

import java.util.HashMap;

public class Copy_List_With_Random_Pointer {
    public class Node {
        int val;
        Node next;
        Node random;

        public Node(int val) {
            this.val = val;
            this.next = null;
            this.random = null;
        }
    }
    public Node copyRandomList(Node head) {
        HashMap<Node, Node> map = new HashMap<>();
        Node cur = head;

        while(cur != null) {
            map.put(cur, new Node(cur.val));
            cur = cur.next;
        }

        cur = head;

        while(cur != null) {
            Node copy = map.get(cur);
            copy.next = map.get(cur.next);
            copy.random = map.get(cur.random);
            cur = cur.next;
        }

        return map.get(head);
    }
    public static void main(String[] args) {

    }
}

File: 2026-02-03/Copy_List_With_Random_Pointer.py
from typing import Optional
from Linked_List import Linked_List, ListNode

class Node:
    def __init__(self, x: int, next: 'Node' = None, random: 'Node' = None):
        self.val = int(x)
        self.next = next
        self.random = random

class Solution:
    def copyRandomList(self, head: 'Optional[Node]') -> 'Optional[Node]':
        if not head:
            return None

        cur = head
        while cur:
            node = Node(cur.val)
            node.next = cur.random
            cur.random = node
            cur = cur.next

        newHead = head.random
        cur = head
        while cur:
            node = cur.random
            node.random = node.next.random if node.next else None
            cur = cur.next

        cur = head
        while cur:
            node = cur.random
            cur.random = node.next
            node.next = cur.next.random if cur.next else None
            cur = cur.next

        return newHead

